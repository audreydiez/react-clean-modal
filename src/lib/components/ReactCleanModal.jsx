import React, { useState, useEffect } from 'react'
import ReactDOM from 'react-dom'
import './ReactCleanModal.scss'
import iconClose from '../assets/icon_close.svg'
import Spinner from './Spinner'

/**
 * Function component that render modal with multiple optionals params
 * @listens {hide} hide - Handle state from parent for hide the modal component
 * @param {boolean} isVisible - State for modal visibility
 * @param {string} customClass - Optional additional css class for all HTML elements
 * @param {boolean} closeOnOverlayClick - Allow close modal on overlay click (not in the modal box)
 * @param {boolean} closeOnScroll - Allow close modal when user scroll (not in the modal box)
 * @param {boolean} animations - Animate the modal ans spinner opening
 * @param {boolean} closeOnTop - Add a svg button at the top with close function
 * @param {string} ariaLabelledBy - Add ariaLabelledBy HTML attribute to modal and title
 * @param {string} testId - add data-testid attribute to the modal container for tests
 * @param {Array.<Object>} customFooter - Array of custom buttons for the footer of the modal
 * @param {string} customFooterAlign - Align the modal footer button to left, right or center
 * @param {boolean} showSpinner - Show spinner before render modal
 * @param {props} props for the component Modal
 *
 * @returns
 */
const ReactCleanModal = ({
    hide,
    isVisible,
    customClass,
    closeOnOverlayClick = false,
    closeOnScroll = false,
    animations = false,
    closeOnTop = false,
    ariaLabelledBy,
    testId,
    customFooter,
    customFooterAlign,
    showSpinner,
    ...props
}) => {
    const [animationsOnClose, setAnimationsOnClose] = useState(false)

    let tempAnimationsOpen = animations

    /* istanbul ignore next */
    function closeModalEvent(e) {
        // When key press event
        if (e.key === 'Escape') {
            if (isVisible && !animations) {
                hide()
            }
            if (isVisible && animations) {
                setAnimationsOnClose(true)
                setTimeout(function () {
                    setAnimationsOnClose(false)
                    hide()
                }, 250)
            }
        }
        // When ordinary closing event
        if (isVisible && !animations) {
            hide()
        }
        if (isVisible && animations) {
            setAnimationsOnClose(true)
            setTimeout(function () {
                setAnimationsOnClose(false)
                hide()
            }, 250)
        }
    }

    /**
     * Add event listener when component is mount.
     * @callback Remove all event listener when unmount
     * @event closeModalEvent#Wheel
     * @event closeModalEvent#keyup
     */
    useEffect(() => {
        window.addEventListener('keyup', (e) => {
            closeModalEvent(e)
        })

        if (closeOnScroll) {
            window.addEventListener('wheel', closeModalEvent)
        }

        // If Spinner, no need animation for modal overlay
        if (showSpinner) {
            tempAnimationsOpen = false
        }

        return () => {
            window.removeEventListener('wheel', closeModalEvent)
            window.removeEventListener('keyup', closeModalEvent)
        }
    }, [isVisible])

    /**
     * Create the custom footer if optional customFooter button are received
     * @returns {Array} arrayOfButton - ready to display in render
     */
    const createCustomFooter = () => {
        const arrayOfBtn = []
        customFooter.map((btn, key) => {
            arrayOfBtn.push(
                React.createElement(
                    'button',
                    {
                        className: btn.className + ' btn' + key,
                        key: key,
                        onClick: btn.eventHandling ? btn.eventHandling : closeModalEvent
                    },
                    btn.text
                )
            )
        })
        return arrayOfBtn
    }

    return isVisible
        ? ReactDOM.createPortal(
              <div
                  id="modal-root"
                  className={`modal-overlay ${
                      tempAnimationsOpen && showSpinner === undefined ? 'open' : ''
                  } ${animationsOnClose ? 'close' : ''} ${
                      customClass ? 'modal-overlay-' + customClass : ''
                  }`}
                  onClick={closeOnOverlayClick ? closeModalEvent : undefined}>
                  <div
                      role="dialog"
                      aria-modal="true"
                      className={`modal modal-container ${animations ? 'open' : ''} ${
                          customClass ? 'modal-container-' + customClass : ''
                      } `}
                      aria-labelledby={ariaLabelledBy}
                      data-testid={testId}
                      onClick={(e) => e.stopPropagation()}>
                      {props.children}
                      {closeOnTop ? (
                          <div
                              className={`modal-close ${
                                  customClass ? 'modal-close-' + customClass : ''
                              }`}
                              onClick={closeModalEvent}>
                              <img
                                  src={iconClose}
                                  className={`modal-close-icon ${
                                      customClass ? 'modal-close-icon-' + customClass : ''
                                  }`}
                                  alt="Close modal"
                              />
                          </div>
                      ) : (
                          ''
                      )}
                      {customFooter ? (
                          <div
                              className={`modal-footer ${
                                  customFooterAlign ? customFooterAlign : 'center'
                              } ${customClass ? 'modal-footer-' + customClass : ''}`}>
                              {createCustomFooter()}
                          </div>
                      ) : (
                          ''
                      )}
                  </div>
              </div>,
              document.body
          )
        : showSpinner
        ? ReactDOM.createPortal(
              <div
                  role="spinner"
                  className={`modal-overlay spinner-overlay ${animations ? 'open' : ''}`}>
                  <Spinner />
              </div>,
              document.body
          )
        : ''
}

export default ReactCleanModal
