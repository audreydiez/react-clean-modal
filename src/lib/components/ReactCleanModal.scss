.modal-overlay {
    position: fixed;
    z-index: 42;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    overflow: hidden;
    background-color: rgba(0, 0, 0, 0.8);
    opacity: 1;

    display: flex;
    flex-direction: row;

    justify-content: center;
    align-items: center;

    &.open {
        animation: fadeIn ease-out 0.1s forwards;
    }
    &.close {
        animation: fadeOut ease-out 0.2s forwards;
    }

    .modal {
        position: relative;

        &-container {
            opacity: 1;
            background-color: white;
            width: 60%;
            height: auto;
            border-radius: 1rem;
            padding: 2rem;
            box-shadow: rgba(0, 0, 0, 0.25) 0px 54px 55px, rgba(0, 0, 0, 0.12) 0px -12px 30px,
                rgba(0, 0, 0, 0.12) 0px 4px 6px, rgba(0, 0, 0, 0.17) 0px 12px 13px,
                rgba(0, 0, 0, 0.09) 0px -3px 5px;

            &.open {
                animation: ShowUp ease-out 400ms forwards;
            }
        }
        &-footer {
            position: relative;
            width: 80%;

            display: flex;
            flex-direction: row;
            flex-wrap: wrap;

            & > button.btn0 {
                order: 1;
            }
            & > button.btn1 {
                order: 2;
            }
            & > button.btn2 {
                order: 3;
            }
            & > button.btn3 {
                order: 4;
            }
            & > button.btn4 {
                order: 5;
            }
            @media screen and (max-width: 650px) {
            }
            @media screen and (max-width: 650px) {
                & > button.btn0 {
                    order: 5;
                }
                & > button.btn1 {
                    order: 4;
                }
                & > button.btn2 {
                    order: 3;
                }
                & > button.btn3 {
                    order: 2;
                }
                & > button.btn4 {
                    order: 1;
                }
            }

            &.center {
                margin: 0 auto;
                text-align: center;
            }
            &.left {
                margin: 0 0;
                text-align: left;
            }
            &.right {
                float: right;
                text-align: right;
            }
        }
        &-footer::after {
            content: '';
            position: absolute;
            background-color: #333;
            width: 100%;
            height: 2px;
            top: 0;
            left: 0;
        }
    }
    .modal-close {
        z-index: 43;
        position: absolute;
        top: -35px;
        right: 5px;
        width: 30px;
        height: 30px;
        cursor: pointer;
        & > * {
            width: 30px;
            height: 30px;
        }
    }
    .btn-clean-modal,
    .btn-clean-modal:visited {
        min-width: 150px;
        background-color: white;
        color: #333;
        font-weight: 600;
        font-size: 1.1rem;
        border: 2px solid #333;
        padding: 0.5rem 1rem;
        border-radius: 0.5rem;
        transition: 0.3s;
        cursor: pointer;
        margin-top: 2rem;

        &:hover {
            background-color: #333;
            color: #fff;
        }
    }
    .btn-clean-modal:not(:first-child):not(:last-child) {
        margin-right: 1rem;
    }
    .btn-clean-modal:first-child {
        margin-right: 1rem;
    }
}

@keyframes fadeIn {
    0% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
}

@keyframes fadeOut {
    0% {
        opacity: 1;
    }
    100% {
        opacity: 0;
    }
}

@keyframes ShowUp {
    0% {
        opacity: 0;
        transform: translateY(50px);
    }
    50% {
        opacity: 0;
    }
    100% {
        transform: translateY(0px);
        opacity: 1;
    }
}
