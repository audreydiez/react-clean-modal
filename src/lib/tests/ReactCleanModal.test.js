import React from 'react'
import { render } from '@testing-library/react'
import ReactCleanModal from '../components/ReactCleanModal'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, within } from '@testing-library/dom'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import renderer from 'react-test-renderer'

//jest.mock('rc-util/lib/Portal')
import EnzymeToJson from 'enzyme-to-json'
const mockFn = jest.fn()
Enzyme.configure({ adapter: new Adapter() })
const arrayOfBtn = [
    {
        text: 'Close modal',
        className: 'btn-clean-modal-one',
        eventHandling: mockFn
    },
    {
        text: 'Alert me!',
        className: 'btn-clean-modal-two',
        eventHandling: mockFn
    }
]

describe('Portal Children test', () => {
    test('Modal shows the children', () => {
        render(
            <ReactCleanModal isVisible={true} closeOnOverlayClick={true}>
                <div data-testid="test" />
            </ReactCleanModal>
        )

        // get element only in modal-root
        const { getByTestId } = within(document.getElementById('modal-root'))
        expect(getByTestId('test')).toBeInTheDocument()
    })
})

describe('Modal rendering', () => {
    test('Render spinner before modal', () => {
        const wrapper = shallow(<ReactCleanModal isVisible={false} showSpinner={true} />)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.exists('div.spinner-overlay')).toEqual(true)
    })

    test('Render modal with animations', () => {
        const wrapper = shallow(<ReactCleanModal isVisible={true} animations={true} />)
        expect(wrapper).toMatchSnapshot()
        const modalOverlay = wrapper.find('.modal-overlay')
        const modalContainer = wrapper.find('.modal-container')
        expect(modalOverlay.exists('.open')).toEqual(true)
        expect(modalContainer.exists('.open')).toEqual(true)
    })

    test('Render modal with custom button and custom function', () => {
        const mockFn = jest.fn()

        const arrayOfBtn = [
            {
                text: 'Close modal',
                className: 'btn-clean-modal-one',
                eventHandling: mockFn
            },
            {
                text: 'Alert me!',
                className: 'btn-clean-modal-two',
                eventHandling: mockFn
            }
        ]

        const wrapper = shallow(<ReactCleanModal isVisible={true} customFooter={arrayOfBtn} />)
        expect(wrapper).toMatchSnapshot()

        expect(wrapper.exists('.btn-clean-modal-one')).toEqual(true)
        expect(wrapper.exists('.btn-clean-modal-two')).toEqual(true)

        const firstButton = wrapper.find('.btn-clean-modal-one')
        firstButton.simulate('click')
        expect(mockFn).toHaveBeenCalled()
    })

    test('Render modal with aria label attribute', () => {
        const wrapper = shallow(
            <ReactCleanModal isVisible={true} ariaLabelledBy={'dialog1_label'}>
                <h1 aria-describedby="dialog1_label">HAHA</h1>
            </ReactCleanModal>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('div.modal-container').props()).toHaveProperty(
            'aria-labelledby',
            'dialog1_label'
        )
    })
    test('Render modal with data-testid attribute', () => {
        const wrapper = shallow(<ReactCleanModal isVisible={true} testId={'modal-test'} />)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('div.modal-container').props()).toHaveProperty(
            'data-testid',
            'modal-test'
        )
    })
    test('Close modal is called when button click', () => {
        const mockCloseFn = jest.fn()

        const wrapper = shallow(
            <ReactCleanModal isVisible={true} hide={mockCloseFn}>
                <button type="button" className="btn-clean-modal" onClick={mockCloseFn}>
                    Close
                </button>
            </ReactCleanModal>
        )

        const firstButton = wrapper.find('.btn-clean-modal')
        firstButton.simulate('click')
        expect(mockCloseFn).toHaveBeenCalled()
    })
    test('tester fonction interieur ', () => {
        const wrapper = shallow(<ReactCleanModal isVisible={true} closeOnOverlayClick={true} />)

        const overlay = wrapper.find('.modal-overlay')

        //console.log(overlay.debug())

        //overlay.simulate('click')

        //const overlayClicked = wrapper.find('.modal-overlay')

        //expect(overlayClicked).toHaveClass('close')
    })
})

it('renders a snapshot', () => {
    const subject = mount(<ReactCleanModal isVisible={true} closeOnOverlayClick={true} />)

    expect(EnzymeToJson(subject)).toMatchSnapshot()
})
