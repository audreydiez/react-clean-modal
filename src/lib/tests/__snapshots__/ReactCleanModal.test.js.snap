// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot 1`] = `
<ReactCleanModal
  closeOnOverlayClick={true}
  isVisible={true}
>
  <Portal
    containerInfo={
      <body>
        <div
          class="modal-overlay   "
          id="modal-root"
        >
          <div
            aria-modal="true"
            class="modal modal-container   "
            role="dialog"
          >
            
            
          </div>
        </div>
      </body>
    }
  >
    <div
      className="modal-overlay   "
      id="modal-root"
      onClick={[Function]}
    >
      <div
        aria-modal="true"
        className="modal modal-container   "
        onClick={[Function]}
        role="dialog"
      />
    </div>
  </Portal>
</ReactCleanModal>
`;
